v3/mathjax3-ts/core/MathDocument.ts

```diff
@@ -0,0 +1,401 @@
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+import {userOptions, defaultOptions, OptionList} from '../util/Options.js';
+import {InputJax, AbstractInputJax} from './InputJax.js';
+import {OutputJax, AbstractOutputJax} from './OutputJax.js';
+import {MathList, MathListClass, AbstractMathList} from './MathList.js';
+import {MathItem, AbstractMathItem} from './MathItem.js';
+
+/*****************************************************************/
+/*
+ *  The MathDocument interface
+ *
+ *  The MathDocument is created by MathJax.Document() and holds the
+ *  document, the math found in it, and so on.  The methods of the
+ *  MathDocument all return the MathDocument itself, so you can
+ *  chain the method calls.  E.g.,
+ *
+ *    const html = MathJax.Document('<html>...</html>');
+ *    html.findMath()
+ *        .compile()
+ *        .getMetrics()
+ *        .typeset()
+ *        .addEventHandlers()
+ *        .updateDocument();
+ *
+ *  The MathDocument is the main interface for page authors to
+ *  interact with MathJax.
+ */
+
+export interface MathDocument {
+    /*
+     * The document being processed (e.g., DOM document, or Markdown string)
+     */
+    document: any;
+
+    /*
+     * The kind of MathDocument (e.g., "HTML")
+     */
+    kind: string;
+
+    /*
+     * The options for the document
+     */
+    options: OptionList;
+
+    /*
+     * The list of MathItems found in this page
+     */
+    math: MathList;
+
+    /*
+     * This object tracks what operations have been performed, so that (when
+     *  asynchronous operations are used), the ones that have already been
+     *  completed won't be performed again.
+     */
+    processed: {[name: string]: boolean};
+
+    /*
+     * An array of input jax to run on the document
+     */
+    inputJax: InputJax[];
+
+    /*
+     * The output jax to use for the document
+     */
+    outputJax: OutputJax;
+
+    /*
+     * Locates the math in the document and constructs the MathList
+     *  for the document.
+     *
+     * @param{OptionList} options  The options for locating the math
+     * @return{MathDocument}       The math document instance
+     */
+    findMath(options?: OptionList): MathDocument;
+
+    /*
+     * Calls the input jax to process the MathItems in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    compile(): MathDocument;
+
+    /*
+     * Gets the metric information for the MathItems
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    getMetrics(): MathDocument;
+
+    /*
+     * Calls the output jax to process the compiled math in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    typeset(): MathDocument;
+
+    /*
+     * Add any event handlers to the typeset math
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    addEventHandlers(): MathDocument;
```

Comment:
But it supposes that we are in some sort of live DOM structure, where events actually exist.
Actually 
> The result of the OutputJax is HTML nodes, or SVG nodes, or MathML nodes, regardless of the 
> document type, and it may be the the typeset math is not actually inserted back into the original 
> document. 
sounds to me that event handling should be separated from the document.
Is it that you are worried that expressions would need to be processed twice?